### ISSUE TITLE
Digital Scale Bar

### ISSUE THREAD

**Issue by afong3** [temporalecologylab/TreeRings#28]
A very necessary step to calculating the physical distance between tree rings is having a scale bar to know the dimensions of each pixel in the image. While it has been normal for other systems to have a ruler or equivalent placed next to a core while imaging it, this is quite difficult with this system.  While this does work well to show scale, it unnecessarily blocks a portion of the cookie. 

My suggestion is to use the generated DPI (Dots Per Inch) metadata from the machine instead of a physical scale bar. To show that the DPI calculations from the machine are reasonably accurate, I have created a digital scale bar from the DPI value which is then overlaid onto a physical scale bar. The physical scale bar is made of a grid of 1mm x 1mm squares. Each resolution has a unique DPI metadata value which was used to create the scale bar. 

Below are a few resolutions (which resulted in small enough files to display in this MD rendering). Although the same logic was used to create digital scale bars for every resolution successfully. 

**5 Percent Resolution:**
![mosaic_5per_scale](https://github.com/user-attachments/assets/507b9c01-262b-49fc-870d-f115393316b2)

**10 Percent Resolution:**
![mosaic_10per_scale](https://github.com/user-attachments/assets/bda6a5b2-2668-4db0-91ad-803c21d66a9c)

**20 Percent Resolution:**
![mosaic_20per_scale](https://github.com/user-attachments/assets/b1286db1-ac4a-4616-8be1-4afadabdc00f)

**Comment by lizzieinvancouver** [comment]
@afong3 Okay, so just to confirm -- these are images with a little scale piece of paper plopped on the cookie? If so, then I think you have convinced me. You should just save some of these photos and your argument here for a supplement to a paper or such.

**Comment by afong3** [comment]
Yup, that's exactly what it is. I've uploaded the code to GitHub and added a set of the stitched images to the Sync vault.

**Comment by afong3** [comment]
Upon further examining the DPI, I'm realizing that it is very difficult to say with sub millimeter confidence that DPI is correct. It was suggested from the beginning to explore adding a physical scale bar to the samples. I assumed that some software would be used to then zoom in on that scale bar and click where the datum is and the distance to a known value. This would be enough information to calculate a DPI value. 

The problem with the current method is that it is calculated from the dimensions of the field of view of one individual frame and the ruler used only has 1mm tick marks but we need to have sub millimeter resolution. Either we need to get a ruler that has sub millimeter tick marks or find some other level of acceptable DPI error. I think we should probably discuss in person though.

From this image below, the scale bar is seemingly just 1cm printed from a standard printer. So in my opinion we are already more accurate than this scale bar but this hasn't been discussed to a conclusion yet.

![image](https://github.com/user-attachments/assets/9cfa76cd-15db-42ef-a028-5fec8a9538ef)

**Comment by lizzieinvancouver** [comment]
@afong3 Happy to discuss in person! You can often pay for a microscope to have a special tick bar so I wonder if we can also pay (perhaps quite a bit) for a fancier measuring device to put on the samples. Maybe you could ask @wangxm-forest to search for such an item at places that sell microscropes?

**Comment by afong3** [comment]
@lizzieinvancouver I think a fancy measuring device would be great. I'll reach out to @wangxm-forest.

**Comment by afong3** [comment]
I'm using this comment as storage for different microscope slide rulers:

| Retailer | Item | Price |
| --- | --- | --- |
| Fischer Scientific | [Stage micrometer calibration ruler](https://www.fishersci.ca/shop/products/horizontal-stage-microme-1ea/507532911#?keyword=calibration%20slide) | $427.26 |

**Comment by lizzieinvancouver** [comment]
@afong3 I like the one that appears to go with a microscrope (the first one). Could you ask @wangxm-forest if she can help us order one (or something similar)? She has been ordering from ThermoFisher I think so should know how. It could go on PAKF if we order soon and we should get a discount so we don't pay so much.

**Comment by afong3** [comment]
DPI measurements have been taken using the slide calibration ruler from Fisher Scientific. Max DPI for the lens set at 2.5x was 21150 dpi. 

## Methods 

1. Wrote a [calibration Python tool](https://github.com/temporalecologylab/TreeRings/blob/main/code/measure_dpi/measure_dpi.py) to calculate DPI from user-identified slide tick marks and the distance between them. 
2. Measured DPI in 6 regions of the image. The idea was to emphasize measurements in the edges of the image, which would be the points of highest lens distortion. In each region distances of tick marks were recorded - tick mark locations were found by zooming to the sub-pixel level and were clicked with a mouse. 
3. In each region, 5 samples were taken using different tick marks in the slide each measurement.
4. Additionally, slides were aligned vertically and horizontally to verify they both converged on the same value.
5. Measurements can be found [here](https://docs.google.com/spreadsheets/d/1iVwj_XTiZ_dKdIyzy0bgV7yw7yNLceb2lGzlhKpOcWQ/edit?gid=0#gid=0).

![Sample Locations](https://github.com/user-attachments/assets/9fc89e77-6d3e-49d4-b8bb-cf8c0b5403f0)

### DPI Conservation in Image Resizing
- I wanted to test if image resizing had a linear relationship with the percentage downscale that is applied.
- I repeated the methods on an image that was resized to 30% of its original size. Measurements were consistent to 30% of the max dpi's value. 

### Implementing into System
- Work is now being put to integrating this information into the system.
